src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt

```diff
@@ -0,0 +1,336 @@
+        .double()
+        .default(1.0)
+        .validate { require(it in 0.5..1.0) {"min-coverage must be between 0.5 and 1.0"} }
+
+    val likelyAncestorFile by option(help="If useLikelyAncestors is true, a record of the ancestors used for each sample will be written to this file.")
+        .default("")
+
+    val threads by option(help = "number of threads used to find paths.").int().default(3)
+
+
+    private val myLogger = LogManager.getLogger(DiploidPathFinding::class.java)
+
+    private fun buildHaplotypeGraph(): HaplotypeGraph {
+        val listOfHvcfFilenames = File(hvcfDir).listFiles().filter { it.name.contains(".h.vcf") }.map { it.path }
+        return HaplotypeGraph(listOfHvcfFilenames)
+    }
+
+    override fun run() {
+
+        val samplesToReadMappingFiles = processKeyFile()
+        processReadMappings(samplesToReadMappingFiles)
+    }
+
+    /**
+     * Process the keyfile. Reads the keyfile. For each sample in the keyfile, imputes a path and writes
+     * it to a .h.vcf file and, optionally, appends the likely parents to a file.
+     */
+    private fun processKeyFile(): Map<String, List<String>> {
+        //processes the key file to produce a map of sample name -> list of read mapping files for that sample
+        myLogger.info("Processing key file $pathKeyfile")
+        val keyFileLines = getBufferedReader(pathKeyfile).use { it.readLines()}
+            .map { it.split("\t") }
+            .filter { it.size > 1 }
+        val headerList = keyFileLines[0]
+
+        val sampleNameIsFirstColumn = headerList[0].equals("SampleName", true)
+        val readMappingIsSecondColumn =  headerList[1].equals("ReadMappingFiles", true)
+        require(sampleNameIsFirstColumn && readMappingIsSecondColumn) {"The first column heading of the keyfile " +
+                "must be SampleName and the second column heading must be ReadMappingFiles."}
+
+        //check for duplicate sample names (not allowed)
+        val sampleNameSet = keyFileLines.drop(1).map { it[0] }.toSet()
+        check(sampleNameSet.size == keyFileLines.size - 1) {"key file contains duplicate sample names"}
+
+        //map of sampleName -> list of read mapping file names
+        val sampleToFiles = keyFileLines.drop(1).associate { Pair(it[0], it[1]) }
+            .mapValues { it.value.split(",").map{filename -> filename.trim()} }
+
+        return checkForExistingOutput(sampleToFiles)
+    }
+
+    /**
+     * Takes a map of samples to file list as input. If the output directory contains an hvcf file for any sample,
+     * that sample will be removed from the map and the trimmed map will be returned.
+     */
+    private fun checkForExistingOutput(sampleToFiles: Map<String, List<String>>): Map<String, List<String>> {
+        val existingSampleFiles = File(outputDir).listFiles()
+        val existingSamples = existingSampleFiles.map {file ->
+            file.name.substringBefore(".h.vcf", "$$$")
```

Comment:
Should you just use an empty string here?  Or rather filter any name missing the .h.vcf before then just do a substringBefore without the second parameter?